packages/snaps-controllers/src/cronjob/CronjobController.ts

```diff
@@ -91,529 +87,420 @@ export const DAILY_TIMEOUT = inMilliseconds(24, Duration.Hour);
 
-  /**
-   * Execute job.
-   *
-   * @param job - Cronjob specification.
-   */
-  async #executeCronjob(job: Cronjob) {
-    this.#updateJobLastRunState(job.id, Date.now());
-    await this.messagingSystem.call('SnapController:handleRequest', {
-      snapId: job.snapId,
-      origin: 'metamask',
-      handler: HandlerType.OnCronjob,
-      request: job.request,
-    });
+    // Cancel all timers and clear the map.
+    this.#timers.forEach((timer) => timer.cancel());
+    this.#timers.clear();
   }
 
   /**
-   * Schedule a background event.
-   *
-   * @param backgroundEventWithoutId - Background event.
-   * @returns An id representing the background event.
+   * Start the daily timer that will reschedule events every 24 hours.
    */
-  scheduleBackgroundEvent(
-    backgroundEventWithoutId: Omit<BackgroundEvent, 'id' | 'scheduledAt'>,
-  ) {
-    // Remove millisecond precision and convert to UTC.
-    const scheduledAt = DateTime.fromJSDate(new Date())
-      .toUTC()
-      .startOf('second')
-      .toISO({
-        suppressMilliseconds: true,
-      });
-
-    assert(scheduledAt);
-
-    const event = {
-      ...backgroundEventWithoutId,
-      id: nanoid(),
-      scheduledAt,
-    };
-
-    this.#setUpBackgroundEvent(event);
-
-    return event.id;
+  #start() {
+    this.#dailyTimer.start(() => this.#reschedule());
   }
 
   /**
-   * Cancel a background event.
+   * Get the next execution date for a given event and start a timer for it.
    *
-   * @param origin - The origin making the cancel call.
-   * @param id - The id of the background event to cancel.
-   * @throws If the event does not exist.
+   * @param event - The event to schedule.
    */
-  cancelBackgroundEvent(origin: string, id: string) {
-    assert(
-      this.state.events[id],
-      `A background event with the id of "${id}" does not exist.`,
-    );
-
-    assert(
-      this.state.events[id].snapId === origin,
-      'Only the origin that scheduled this event can cancel it.',
-    );
-
-    const timer = this.#timers.get(id);
-    timer?.cancel();
-    this.#timers.delete(id);
-    this.#snapIds.delete(id);
+  #schedule(event: InternalBackgroundEvent) {
+    const date = getExecutionDate(event.schedule);
     this.update((state) => {
-      delete state.events[id];
+      state.events[event.id].date = date;
+    });
+
+    this.#startTimer({
+      ...event,
+      date,
     });
   }
 
   /**
-   * A helper function to handle setup of the background event.
+   * Set up and start a timer for the given event.
    *
-   * @param event - A background event.
+   * @param event - The event to schedule.
+   * @throws If the event is scheduled in the past.
    */
-  #setUpBackgroundEvent(event: BackgroundEvent) {
-    const date = new Date(event.date);
-    const now = new Date();
-    const ms = date.getTime() - now.getTime();
-
+  #startTimer(event: InternalBackgroundEvent) {
+    const ms = DateTime.fromISO(event.date).toMillis() - Date.now();
```

Comment:
Should this have `setZone: true`?

Reply:
I don't think so, since `Date.now()` is based on UTC as well. But timezones are confusing so I may be wrong 😅

Reply:
I think `fromISO` uses the local timezone by default if you don't, but I could be wrong